pkg/serializer/internal/metrics/events.go

```diff
@@ -280,3 +292,240 @@ func (events Events) CreateMarshalersBySourceType() []marshaler.StreamJSONMarsha
+		if err != nil {
+			return err
+		}
+
+		if e.eventsInPayload {
+			tlmEventsPayloads.Inc()
+			e.payloads = append(e.payloads, transaction.NewBytesPayload(payload, 0))
+		} else {
+			tlmEventsEmptyPayloads.Inc()
+		}
+	}
+
+	e.bufferContext.CompressorInput.Reset()
+	e.bufferContext.CompressorOutput.Reset()
+
+	return nil
+}
+
+func (e *eventsMarshaler2) writeEvent(ev *event.Event) {
+	s := e.stream
+	s.WriteObjectStart()
+	s.WriteObjectField("msg_title")
+	s.WriteString(ev.Title)
+	s.WriteMore()
+	s.WriteObjectField("msg_text")
+	s.WriteString(ev.Text)
+	s.WriteMore()
+	s.WriteObjectField("timestamp")
+	s.WriteInt64(ev.Ts)
+	if ev.Priority != "" {
+		s.WriteMore()
+		s.WriteObjectField("priority")
+		s.WriteString(string(ev.Priority))
+	}
+	s.WriteMore()
+	s.WriteObjectField("host")
+	s.WriteString(ev.Host)
+	if len(ev.Tags) > 0 {
+		s.WriteMore()
+		s.WriteObjectField("tags")
+		s.WriteArrayStart()
+		for i, t := range ev.Tags {
+			if i > 0 {
+				s.WriteMore()
+			}
+			s.WriteString(t)
+		}
+		s.WriteArrayEnd()
+	}
+	if ev.AlertType != "" {
+		s.WriteMore()
+		s.WriteObjectField("alert_type")
+		s.WriteString(string(ev.AlertType))
+	}
+	if ev.AggregationKey != "" {
+		s.WriteMore()
+		s.WriteObjectField("aggregation_key")
+		s.WriteString(ev.AggregationKey)
+	}
+	if ev.SourceTypeName != "" {
+		s.WriteMore()
+		s.WriteObjectField("source_type_name")
+		s.WriteString(ev.SourceTypeName)
+	}
+	if ev.EventType != "" {
+		s.WriteMore()
+		s.WriteObjectField("event_type")
+		s.WriteString(ev.EventType)
+	}
+	s.WriteObjectEnd()
+}
+
+// returns true if item was processed or dropped, false if we need to try again
+func (e *eventsMarshaler2) writeItem(ev *event.Event) (bool, error) {
+	s := e.stream
+	s.Reset(nil)
+
+	if !e.eventsInPayload || e.sourceTypeName != ev.SourceTypeName {
+		if e.eventsInPayload {
+			s.WriteArrayEnd()
+			s.WriteMore()
+		}
+		if ev.SourceTypeName != "" {
+			s.WriteObjectField(ev.SourceTypeName)
+		} else {
+			s.WriteObjectField("api")
+		}
+		s.WriteArrayStart()
+	} else {
+		s.WriteMore()
+	}
+
+	e.writeEvent(ev)
+
+	err := e.compressor.AddItem(s.Buffer())
+	switch err {
+	case nil:
+		e.sourceTypeName = ev.SourceTypeName
+		e.eventsInPayload = true
+		tlmEventsSent.Inc()
+		return true, nil
+	case stream.ErrPayloadFull:
+		if err := e.flushPayload(); err != nil {
+			return false, err
+		}
+		e.eventsInPayload = false
+		return false, nil
+	default:
+		tlmEventsDropped.Inc()
+		e.logger.Warnf("Dropping event: title=%q text=%q source_type_name=%q: %v", ev.Title, ev.Text, ev.SourceTypeName, err)
+		return true, nil
+	}
+}
+
+func (e *eventsMarshaler2) marshal() ([]*transaction.BytesPayload, error) {
+	if err := e.flushPayload(); err != nil {
+		return nil, err
+	}
+	for _, ev := range e.events {
+		for {
+			ok, err := e.writeItem(ev)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				break
+			}
+		}
+	}
+	if err := e.closePayload(); err != nil {
+		return nil, err
+	}
+
+	return e.payloads, nil
```

Comment:
Can't you
```
return e.closePayload()
```

You really don't want to return `e.payloads` in case of err?

Reply:
> `return e.payloads, e.closePayload()`

Go evaluates left to right, so this would loose last payload if `closePayload` needs to re-allocate `e.payloads`.